library/src/blas1/rocblas_copy.hpp

```diff
@@ -0,0 +1,75 @@
+/* ************************************************************************
+ * Copyright 2016-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#include "handle.h"
+#include "logging.h"
+#include "rocblas.h"
+#include "utility.h"
+
+template <typename T, typename U, typename V>
+__global__ void copy_kernel(rocblas_int    n,
+                            const U        xa,
+                            rocblas_int    offsetx,
+                            rocblas_int    incx,
+                            rocblas_stride stridex,
+                            V              ya,
+                            rocblas_int    offsety,
+                            rocblas_int    incy,
+                            rocblas_stride stridey)
+{
+    ptrdiff_t tid = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    // bound
+    if(tid < n)
+    {
+        const T* x = load_ptr_batch(xa, hipBlockIdx_y, offsetx, stridex);
+        T*       y = load_ptr_batch(ya, hipBlockIdx_y, offsety, stridey);
+
+        if(incx < 0)
+            x -= ptrdiff_t(incx) * (n - 1);
+        if(incy < 0)
+            y -= ptrdiff_t(incy) * (n - 1);
```

Comment:
The adjustments based on `incx`, `incy` should be moved outside of the kernel, to where they were originally (`_impl`). Not sure how this interacts with `load_ptr_batch`, but the way it's written right now, will be much slower.

Reply:
This doesn't really seem possible with the way we've chosen to combine the non-batched/batched/strided_batched into a single template. Although, I agree that the non-batched version will be slower. How do you suggest we tackle this?

Reply:
I would think going back to the original kernel is valid option for the base case.   Reusing generic kernels isn't a great improvement when there is so little code that is duplicated anyway and making it much more complicated with unused parameters outweighs the potential maintenance improvement in my opinion.

Reply:
> This doesn't really seem possible with the way we've chosen to combine the non-batched/batched/strided_batched into a single template. Although, I agree that the non-batched version will be slower. How do you suggest we tackle this?

I thought we were only merging the batched kernels, and that the non-batched kernels would stay separate.

There should be a way to pass in the "adjusted" x, y, etc. rather than calculating them **_at every single element._**

This should be a separate JIRA task I'm willing to take on -- to refactor all of the rocBLAS kernels' templates so that the overhead is avoided.

It can be done with template specializations (or, in C++17, with `constexpr if`).

It might be as simple as adding a `bool` template parameter which specifies whether to "adjust" the `x` and `y` pointers. If `true`, it's done the way it's done now. If `false`, it would be done once before the GPU kernel.

Reply:
The few functions that have been combined to a single template are combining all three, this seems to be the first of which that handle the negative increments. GER and GEMV still need to be refactored to a single template.

If you would like to take on this task, how do you suggest we proceed with this PR? Should I revert back to the original kernel for now on the non batched case?

From what I can tell only the batched version needs to be adjusted at the kernel level.

Reply:
You can look at swap which supports negative increments.  I reuse an inline kernel for calculation logic but the ptr_loads are only in the batched kernels

Reply:
Swap has not combined any templates. Should I revert back to this style for copy and dot?

Reply:
For swap all 3 variants call the same inline template inside their kernels.  It is reuse without wasted parameter overhead.   If you want it to run with the same speed as the original non-batched and also share the code that is an option.